core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java

```diff
@@ -174,11 +178,22 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
             InternalNode chosenNode = null;
             long minWeight = Long.MAX_VALUE;
 
-            for (InternalNode node : candidateNodes) {
-                long totalSplitsWeight = assignmentStats.getTotalSplitsWeight(node);
-                if (totalSplitsWeight < minWeight && totalSplitsWeight < maxSplitsWeightPerNode && assignmentStats.getUnacknowledgedSplitCountForStage(node) < maxUnacknowledgedSplitsPerTask) {
-                    chosenNode = node;
-                    minWeight = totalSplitsWeight;
+            if (stageSplitsBalancing) {
+                for (InternalNode node : getFreeNodesForStage(assignmentStats, candidateNodes)) {
+                    long queuedWeight = assignmentStats.getQueuedSplitsWeightForStage(node);
+                    if (queuedWeight < minWeight) {
```

Comment:
nit: `<=` or assert that `chosenNode` is not null at the end of loop

Reply:
`chosenNode` can be null if `getFreeNodesForStage` is empty